rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -359,47 +266,63 @@ rmw_ret_t ClientData::take_response(
+
+  if (!reply.is_ok()) {
     RMW_SET_ERROR_MSG("invalid reply sample");
     return RMW_RET_ERROR;
   }
-  const z_loaned_sample_t * sample = latest_reply->get_sample().value();
-
-  // Object that manages the raw buffer
-  z_owned_slice_t payload;
-  z_bytes_to_slice(z_sample_payload(sample), &payload);
-  eprosima::fastcdr::FastBuffer fastbuffer(
-    reinterpret_cast<char *>(const_cast<uint8_t *>(z_slice_data(z_loan(payload)))),
-    z_slice_len(z_loan(payload)));
 
-  // Object that serializes the data
-  rmw_zenoh_cpp::Cdr deser(fastbuffer);
-  if (!response_type_support_->deserialize_ros_message(
-      deser.get_cdr(),
-      ros_response,
-      response_type_support_impl_))
-  {
-    RMW_SET_ERROR_MSG("could not deserialize ROS response");
-    return RMW_RET_ERROR;
-  }
+  const zenoh::Sample & sample = reply.get_ok();
 
-  // Fill in the request_header
-  AttachmentData attachment(z_sample_attachment(sample));
-  request_header->request_id.sequence_number = attachment.sequence_number();
-  if (request_header->request_id.sequence_number < 0) {
-    RMW_SET_ERROR_MSG("Failed to get sequence_number from client call attachment");
-    return RMW_RET_ERROR;
-  }
-  request_header->source_timestamp = attachment.source_timestamp();
-  if (request_header->source_timestamp < 0) {
-    RMW_SET_ERROR_MSG("Failed to get source_timestamp from client call attachment");
+  // Object that manages the raw buffer
+  auto payload = sample.get_payload().as_vector();
+  if (payload.size() > 0) {
+    eprosima::fastcdr::FastBuffer fastbuffer(
+      reinterpret_cast<char *>(const_cast<uint8_t *>(payload.data())), payload.size());
+
+    // Object that serializes the data
+    rmw_zenoh_cpp::Cdr deser(fastbuffer);
+    if (!response_type_support_->deserialize_ros_message(
+        deser.get_cdr(),
+        ros_response,
+        response_type_support_impl_))
+    {
+      RMW_SET_ERROR_MSG("could not deserialize ROS response");
+      return RMW_RET_ERROR;
+    }
+
+    // Fill in the request_header
+    if (!sample.get_attachment().has_value()) {
+      RMW_ZENOH_LOG_DEBUG_NAMED(
+        "rmw_zenoh_cpp",
+        "ClientData take_request attachment is empty");
+      return RMW_RET_ERROR;
+    }
+    rmw_zenoh_cpp::AttachmentData attachment(std::move(sample.get_attachment().value().get()));
+    request_header->request_id.sequence_number = attachment.sequence_number();
+    if (request_header->request_id.sequence_number < 0) {
+      RMW_SET_ERROR_MSG("Failed to get sequence_number from client call attachment");
+      return RMW_RET_ERROR;
+    }
+    request_header->source_timestamp = attachment.source_timestamp();
+    if (request_header->source_timestamp < 0) {
+      RMW_SET_ERROR_MSG("Failed to get source_timestamp from client call attachment");
+      return RMW_RET_ERROR;
+    }
+    memcpy(
+      request_header->request_id.writer_guid,
+      attachment.copy_gid().data(),
+      RMW_GID_STORAGE_SIZE);
```

Comment:
I'm just noticing that this is doing two copies; once when `attachment.copy_gid()` is called, and then again during the memcpy.  I think we should reduce this back to one copy.